### PR TITLE
Add Jolokia agent in proxy mode

### PR DESCRIPTION
**Pull request by whale-shark**
Modified the object of sending to support Jolokia proxy mode.

### CODE REVIEW

**Comment by ruflin** [comment]
@whale-shark Thanks a lot for opening this PR. Could you make Houndci happy and add an entry to the CHANGELOG.asciidoc file? I'm pretty sure this change will make quite a few people happy.

I was wondering if there is a good way to have automated tests for this?

**Comment by whale-shark** [comment]
@ruflin I plan to carry out CI correspondence and document addition on the weekend.

I've just started Go language, and I'm also studying about testing.
Jolokia's test did not work for both `make test-module [no test files]` and `go test jmx_integration_test.go` from the beginning.
I also want to do a good test and good automated test.

Absolutely test is necessary?

**Comment by ruflin** [comment]
I wonder if jolokia can be run in proxy mode also if it's run on the same machine? Best would be to have also a docker container with the full setup we then can connect to to test this the same way as we do it now. Perhaps we could even have 2 instances of Jolokia running, one in proxy mode and one not?

The main reason I'm asking for tests is because in case it breaks I'm looking for an easy way to have a setup to also test it manually. And there is obviously the benefit of the automated way of knowing when it breaks.

Please post any questions you have for the setup or beats here, more then happy to help you get going.

**Comment by ruflin** [comment]
@whale-shark Any update here? Would be great to move this forward.

**Comment by whale-shark** [comment]
Sorry for late reply.
I'm busy with my job and I've not been able enough time to create a test.

I am considering adding an instance of docker to connect to remote JMX.

Please wait just a little longer.

**Comment by whale-shark** [comment]
I am sorry to be late.
Unit test and integration test added.

I changed Dockerfile.
In the original configuration, it can not be launched in proxy mode.
So I deployed jolokia's war to Tomcat to work in proxy mode.
Also, we added a new Tomcat to be connected in proxy mode.

**Comment by ruflin** [comment]
@whale-shark Any chance you could rebase this on the most recent version of master?

File: CHANGELOG.asciidoc [hunk 1]
```diff
@@ -519,6 +519,13 @@ https://github.com/elastic/beats/compare/v6.0.1...v6.1.0[View commits]
 - Add experimental Docker autodiscover functionality. {pull}5245[5245]
 - Add Windows service metricset in the windows module. {pull}5332[5332]
 - Update gosigar to v0.6.0. {pull}5775[5775]
+- Add experimental system/raid metricset. {pull}5642[5642]
+- Add a dashboard for the Nginx module. {pull}5991[5991]
+- Add experimental mongodb/collstats metricset. {pull}5852[5852]
+- Update the MySQL dashboard to use the Time Series Visual Builder. {pull}5996[5996]
+- Add experimental uwsgi module. {pull}6006[6006]
+- Docker and Kubernetes modules are now GA, instead of Beta. {pull}6105[6105]
+- Add Jolokia agent in proxy mode. {pull}6475[6475]
```
Comment (jsoriano) [inline CHANGELOG.asciidoc]: @whale-shark something weird happened with the changelog after updating the branch, could you remove the added lines that are not related with your change?

**Comment by jsoriano** [comment]
jenkins, test it please

**Comment by jsoriano** [comment]
jenkins, test this please

**Comment by whale-shark** [comment]
@jsoriano Thank you for pointing out

I revert to latest branch and added my fix line.

**Comment by jsoriano** [comment]
jenkins, test it please

**Comment by jsoriano** [comment]
jenkins, test this

File: metricbeat/module/jolokia/_meta/Dockerfile [hunk 2]
```diff
@@ -1,11 +1,39 @@
...
-CMD env CATALINA_OPTS=$(jolokia_opts) /opt/${TC}/bin/catalina.sh run
+# Prepare a server where jolokia runs in proxy mode
+RUN wget http://archive.apache.org/dist/tomcat/tomcat-7/v${TOMCAT_VERSION}/bin/${TC}.tar.gz;\
+    tar xzf ${TC}.tar.gz -C /usr;\
+    sed -i -e 's/Connector port="8080"/Connector port="8778"/g' /usr/${TC}/conf/server.xml;\
+    wget http://central.maven.org/maven2/org/jolokia/jolokia-war/${JOLOKIA_VERSION}/jolokia-war-${JOLOKIA_VERSION}.war -O /usr/${TC}/webapps/jolokia.war
+
+# Set up a server to be connected in proxy mode
+RUN wget http://archive.apache.org/dist/tomcat/tomcat-9/v${PROXT_TARGET_TOMCAT_VERSION}/bin/${PX_TC}.tar.gz;\
+    tar xzf ${PX_TC}.tar.gz -C /usr;\
```
Comment (jsoriano) [inline metricbeat/module/jolokia/_meta/Dockerfile]: Can you also remove these .tar.gz files that are not needed in the final image?

File: metricbeat/module/jolokia/_meta/Dockerfile [hunk 3]
```diff
@@ -1,11 +1,39 @@
...
+    echo "controlRole R&D" >> /usr/lib/jvm/java-1.8-openjdk/jre/lib/management/jmxremote.password;\
+    chmod 600 /usr/lib/jvm/java-1.8-openjdk/jre/lib/management/jmxremote.password
+
+# Start tomcat to accept JMX connection and enable jolokia proxy mode
+CMD env CATALINA_OPTS="$(jolokia_opts)\
+    -Dcom.sun.management.jmxremote.port=7091\
+    -Dcom.sun.management.jmxremote.ssl=false\
+    -Dcom.sun.management.jmxremote.authenticate=true"\
+    /usr/${PX_TC}/bin/catalina.sh run & \
+    env CATALINA_OPTS="-Dorg.jolokia.jsr160ProxyEnabled" /usr/${TC}/bin/catalina.sh run
```
Comment (jsoriano) [inline metricbeat/module/jolokia/_meta/Dockerfile]: Would it be possible to run everything in an only tomcat?

File: metricbeat/module/jolokia/_meta/Dockerfile (hunk 2 above)
Comment (whale-shark) [reply metricbeat/module/jolokia/_meta/Dockerfile]: I deleted the unnecessary tar.gz file from the final image.

File: metricbeat/module/jolokia/_meta/Dockerfile (hunk 3 above)
Comment (whale-shark) [reply metricbeat/module/jolokia/_meta/Dockerfile]: In order to check the connection in Proxy mode, Tomcat to be connected is added.
If you use in an only Tomcat, you will not know if you are connected to another Tomcat in the process of connecting in Proxy mode of the integration test.

Comment (jsoriano) [reply metricbeat/module/jolokia/_meta/Dockerfile]: Ok, makes sense, maybe we could think in a two container scenario, but I think we can continue as is.

**Comment by jsoriano** [comment]
jenkins, test this please

File: metricbeat/module/jolokia/_meta/Dockerfile (hunk 3 above)
Comment (whale-shark) [reply metricbeat/module/jolokia/_meta/Dockerfile]: Changed the connecting to another Tomcat to a pattern where jmx password authentication cannot be performed correctly.
Fix to be possible to run everything in an only Tomcat.

**Comment by jsoriano** [comment]
jenkins, test this please

**Comment by whale-shark** [comment]
@jsoriano  Fixed a file conflicting with the latest source.

**Comment by jsoriano** [comment]
jenkins, test this again, please

**Comment by jsoriano** [comment]
Failing tests are not related.
